### PR TITLE
full review of KPI implementation

### PR DESCRIPTION
**Pull request by tomkralidis**
This PR implements the full review of the KPI implementation as described in #69 as a first pass.  Notes:

- diectory `KPI/` is renamed to `kpi/`
- KPIs are no longer alphabetical, but numeric
- full link check performed
- added https://github.com/wmo-im/wcmp/tree/issue-69/kpi with instructions on how to build and check for stale links
- added overview/table of contents, terms/acronyms
- various editorial updates
- some areas for clarification marked with `TODO`
- all scoring totals bolded
- HTTP links converted to HTTPS where possible

NOTE: for better or worse I normalized all WMO and ISO codelists (there a few variations) as follows:
- WMO codelists: https://wis.wmo.int/2012/codelists/WMOCodeLists.xml
- ISO codelists: https://standards.iso.org/iso/19139/resources/gmxCodelists.xml

..which will make for an easy find/replace job as required.

Temporary HTML and PDF outputs available below:
  - PDF: https://github.com/wmo-im/wcmp/blob/issue-69/kpi/wcmp-1.3-kpi.pdf
  - HTML: http://htmlpreview.github.io/?https://github.com/wmo-im/wcmp/blob/issue-69/kpi/wcmp-1.3-kpi.html

@wmo-im/tt-wismd please review and provide feedback in this issue.  As well, feel free to make commits to this branch (`issue-69`).

Following this feedback, we can circulate to TT-WIGOSMD/others accordingly.

### CODE REVIEW

File: kpi/002.adoc [hunk 1]
```diff
@@ -1,12 +1,12 @@
...
+== KPI-{counter:kpi}: Good quality title
 
 === Status
 
 Approved
 
 === WCMP element(s)
 
-`/gmd:MD_Metadata/gmd:identificationInfo//gmd:citation/gmd:CI_Citation/gmd:title`
+* `/gmd:MD_Metadata/gmd:identificationInfo//gmd:citation/gmd:CI_Citation/gmd:title`
```
Comment (amilan17) [inline kpi/002.adoc:9]: line 13: The title field shall always be populated and ~and~ follow the principles of...

File: kpi/003.adoc [hunk 2]
```diff
@@ -1,12 +1,12 @@
...
 
 === Status
 
-Draft
+Approved
 
 === WCMP element(s)
 
-`/gmd:MD_Metadata/gmd:identificationInfo//gmd:abstract`
+* `/gmd:MD_Metadata/gmd:identificationInfo//gmd:abstract`
```
Comment (amilan17) [inline kpi/003.adoc:9]: line 22:  "~Fulsome~ Complete and meaningful abstract information allows..."

File: kpi/003.adoc [hunk 3]
```diff
@@ -42,7 +42,7 @@ in support of data access, visualization and exploitation.
 |-1
 |===
 
-Total possible score: 3 (100%)
+*Total possible score: 3 (100%)*
```
Comment (amilan17) [inline kpi/003.adoc:45]: line 41 and 42: "abstract contains bulletin template | -1 "

I recommend making this a positive test, for example: "abstract does not contain a bulletin template | +1" 

and then the total score is 4....

File: kpi/003.adoc [hunk 4]
```diff
@@ -52,44 +52,45 @@ The abstract should provide a clear and concise statement that enables the
 reader to understand the content of the dataset. For guidance when completing
 the abstract, consider the following recommendations:
 
-. State what the 'things' are that are recorded
-. State the key aspects recorded about these things
-. State what form the data takes
-. State any other limiting information, such as time period of validity of the
+* State what the 'things' are that are recorded
```
Comment (amilan17) [inline kpi/003.adoc:55]: "State ~what~ the 'things' ~are~ that are recorded"

File: kpi/004.adoc [hunk 5]
```diff
@@ -140,8 +136,9 @@ update frequency, `gmd:MD_MaintenanceFrequencyCode` can be set to `asNeeded`:
 |1
 |===
 
-Total possible score: 5 (100%)
+*Total possible score: 5 (100%)*
```
Comment (amilan17) [inline kpi/004.adoc:139]: Lines 123, 132 and 135

I recommend replacing "... is defined" with "is included". 

This is more consistent with the language used in the WIS Manual. e.g. "Requirement 8.2.4: Each WIS discovery metadata record describing geographic data shall **include** the description of ..."

File: kpi/004.adoc [hunk 6]
```diff
@@ -140,8 +136,9 @@ update frequency, `gmd:MD_MaintenanceFrequencyCode` can be set to `asNeeded`:
...
 
 === Guidance to score well on this assessment
 
 Ensure that the Temporal Extent and the Update Frequency is defined in the metadata record.
+
```
Comment (amilan17) [inline kpi/004.adoc:144]: line 143: ".... Frequency is ~defined~ included in the metadata record."

File: kpi/004.adoc [hunk 7]
```diff
@@ -1,13 +1,8 @@
...
-Metadata records should contain the following temporal information:
-
-. Temporal extent: This element describes the period of time that the available product covers
-. Data Update frequency: The temporal frequency at which the data is updated (i.e. every x hours, days, ...)
+== KPI-{counter:kpi}: Temporal information
 
 === Status
 
-Draft
+Approved
```
Comment (amilan17) [inline kpi/004.adoc:5]: add Xpath for //gmd:status in WCMP Elements section.

File: kpi/004.adoc [hunk 8]
```diff
@@ -16,7 +11,8 @@ Draft
 
 === What is being measured
 
-The presence of a correctly defined temporal extent and data update frequency.
+* Temporal extent: This element describes the period of time that the available product covers
+* Data update frequency: The temporal frequency at which the data is updated (i.e. every x hours, days, etc.)
```
Comment (amilan17) [inline kpi/004.adoc:15]: ADD bullet for 

*   Status of dataset

**Review by amilan17 — COMMENTED** [review]
I think we should remove the ids, because id attributes are not required by the schema and there are no best practices for this in our KPI.... 
* <gmd:EX_TemporalExtent id="temporalExtent">
* <gmd:EX_TemporalExtent id="boundingTemporalExtent">

**Comment by josusky** [comment]
> I think we should remove the ids, because id attributes are not required by the schema and there are no best practices for this in our KPI....

I agree with Anna that the "id" attributes are superfluous and somewhat confusing.

File: kpi/002.adoc [hunk 9]
```diff
@@ -71,5 +70,5 @@ metadata record (abstract, keywords, etc.).
 For a satellite product offering one main data parameter, the title will
 typically define which parameter is contained in the product, and from which
 instrument or instrument type it originates. For example, "AMSR-2 Sea Surface
-Temperature” or “SLSTR L1B radiances and brightness temperatures”.footnote:[WCMP Guidance (http://wis.wmo.int/file=3291)]
+Temperature” or “SLSTR L1B radiances and brightness temperatures”.footnote:[WCMP Guidance (https://wis.wmo.int/file=3291)]
```
Comment (martaguti) [inline kpi/002.adoc:73]: The link https://wis.wmo.int/file=3291  points to a draft document in wmosharepoint

File: kpi/004.adoc (hunk 5 above)
Comment (tomkralidis) [reply kpi/004.adoc:139]: Fixed globally.

File: kpi/002.adoc (hunk 9 above)
Comment (tomkralidis) [reply kpi/002.adoc:73]: @amilan17 can you you point to us to the correct URL here?

File: kpi/008.adoc [hunk 10]
```diff
@@ -1,14 +1,17 @@
...
 
-Any element or attribute content with linked information (URLs). Links are
-generally present in `gmd:URL`, `gmd:graphicOverview`, `gmx:Anchor/@xlink:href`,
-and `@codeList` elements/attributes.
+Any element or attribute content with linked information (URLs).
+
+* `//gmd:URL`
+* `//gmd:graphicOverview`
+* `//gmx:Anchor/@xlink:href`
+* `//@codeList` TODO: should we really check code lists link health or treat as a plain old value?
```
Comment (martaguti) [inline kpi/008.adoc:14]: @tomkralidis for discussion,
 I would check

Comment (amilan17) [reply kpi/008.adoc:14]: I think we should at least test for valid URLs....

File: kpi/009.adoc [hunk 11]
```diff
@@ -258,3 +256,4 @@ If the codelist implementation is not clear there should be additional free text
 `gmd:otherConstraints` or `gmd:useLimitation`.
 
 Further guidance on data policy implementation can be found in the link:https://library.wmo.int/doc_num.php?explnum_id=10257[Guide to WMO Information System (WMO No. 1061, Section 5.8.1.10)].
+
```
Comment (amilan17) [inline kpi/009.adoc:259]: Line 253-255: "In addtion to programmatic checks which will provide an score, it should be clear to the user what are the conditions of use for the resource published."

suggested revision

"In addition to programmatic checks which will provide a score, the conditions of use for the resource published should be clear to the user."

File: kpi/010.adoc [hunk 12]
```diff
@@ -1,13 +1,13 @@
...
 Metadata records should contain information regarding how to access the data
 which it is describing.
 
 === Status
 
-Draft
+Approved
 
-=== WMCP element(s)
+=== WCMP element(s)
```
Comment (amilan17) [inline kpi/010.adoc:10]: line 39 in xml example: 

DT/DSI/OP/TEL is not an individual. Not critical, but this is a common BP that I don't want to perpetuate.

File: kpi/012.adoc [hunk 13]
```diff
@@ -1,4 +1,4 @@
-== N. DOI citation
+== KPI-{counter:kpi}: DOI citation
```
Comment (amilan17) [inline kpi/012.adoc:1]: Did we decide to make this a recommendation instead of KPI requirement?

File: kpi/template.adoc [hunk 14]
```diff
@@ -1,20 +1,29 @@
...
 
+Text description of measurement.
+
 === Rationale for measurement
 
+Why should this measurement take place.
+
 === Rules for implementation
 
 - use XPath and pseudo-code (if then else)
```
Comment (amilan17) [inline kpi/template.adoc:22]: I recommend moving Xpath and pseudo-code to the Guidance section.

File: kpi/template.adoc [hunk 15]
```diff
@@ -1,20 +1,29 @@
...
 - provide total possible score
 
+*Total possible score: x (100%)*
+
 === Guidance to score well on this assessment
```
Comment (amilan17) [inline kpi/template.adoc:27]: If we move xpaths, etc to this section. Change title to Guidance and Examples ....

**Review by amilan17 — CHANGES_REQUESTED** [review]

File: kpi/008.adoc (hunk 10 above)
Comment (tomkralidis) [reply kpi/008.adoc:14]: TT-WISMD meeting 2021-02-19: As discussed, we will check for HTTP 200 and not inspect content.

File: kpi/006.adoc [hunk 16]
```diff
@@ -67,30 +66,29 @@ values against https://wis.wmo.int/2012/codelists/WMOCodeLists#MD_KeywordTypeCod
...
+Check for well referenced thesaurus using the Anchor with a resolvable HTTP URL.
 
 ```xml
 <gmd:descriptiveKeywords>
   <gmd:MD_Keywords>
     <gmd:thesaurusName>
       <gmd:CI_Citation>
         <gmd:title>
-          <gmx:Anchor xlink:href="http://codes.wmo.int/grib2/codeflag/_4.2">WMO Codes Registry GRIB edition 2</gmx:Anchor>
+          <gmx:Anchor xlink:href="https://codes.wmo.int/grib2/codeflag/_4.2">WMO Codes Registry GRIB edition 2</gmx:Anchor>
```
[the remark this answers is not in the document]
Comment (tomkralidis) [reply kpi/006.adoc:83]: TT-WISMD meeting 2021-02-19: use WMDR codes (example: http://codes.wmo.int/wmdr/ApplicationArea/_highResolutionNumericalWeatherPrediction)

File: kpi/004.adoc (hunk 6 above)
[the remark this answers is not in the document]
Comment (amilan17) [reply kpi/004.adoc:146]: #90

File: kpi/006.adoc [hunk 17]
```diff
@@ -135,34 +133,35 @@ instead).
...
+To assess how many keywords are defined and provide a total percentage
+based on the number of keywords and individual scoring.
 
 === Guidance to score well on this assessment
 
 A high score will be provided for full referenced term, keyword type and
 thesaurus. Additional recommendations for keywords implementation are found at
-the Guide to WMO Information System (WMO- No. 1061,Section 5.8.1.8)
-https://library.wmo.int/doc_num.php?explnum_id=4004
+the link:https://library.wmo.int/doc_num.php?explnum_id=4004[Guide to WMO Information System (WMO- No. 1061,Section 5.8.1.8)]
```
[the remark this answers is not in the document]
Comment (amilan17) [reply kpi/006.adoc]: #90

**Pull request merged**